torch/distributed/fsdp/_optim_utils.py

```diff
@@ -497,26 +499,11 @@ def _flatten_tensor_optim_state(
         for state_value, shape
         in zip(pos_dim_tensors, unflat_param_shapes)
     ]
-    padding = flat_param.num_padded
```

Comment:
Adding this padding was unnecessary. `flat_param.num_padded` is the number of elements padded for this local rank's shard of the flattened parameter.

`flat_tensor` is the concatenation of all flattened unsharded optimizer state tensors, which means it has the same shape as the flattened unsharded parameter without padding. (See the new assert on L504.)

If we pad `flat_tensor` by any amount between 0 and the total amount of padding for the flattened parameter, then the padded `flat_tensor` will still shard correctly. All it means is that the amount of padding added in `FullyShardedDataParallel._get_shard()` will be less (i.e. we pad some here and pad some in `_get_shard()` instead of padding all in `_get_shard()`).

Because `flat_param.num_padded` satisfies that inequality of being between 0 and the total amount of padding, this did not cause any correctness issues. However, because this is confusing since we pad the unsharded `flat_tensor` by possibly a different amount per rank, I am removing this logic.

Reply:
Sounds good! Does FSDP._get_shard() mention that it handles padding in the documentation so users of the API don't have to (and probably shouldn't) handle padding themselves?

Reply:
We can clarify that the input tensor to `_get_shard()` does not need to be padded and that the method will pad it.